\name{ipcaps}
\alias{ipcaps}
%- Also NEED an '\alias' for EACH other topic documented here.
\title{
%%  ~~function to do ... ~~
A function for unsupervised clustering to capture population structure based on iterative pruning.
}
\description{
%%  ~~ A concise (1-5 lines) description of what the function does. ~~
This version supports ordinal data which can be applied directly to SNP data to identify fine-scale population structure. It was built on the iterative pruning Principal Component Analysis (ipPCA) algorithm [1]. The IPCAPS involves an iterative process using multiple splits based on multivariate Gaussian mixture modeling of principal components and Clustering EM estimation [2]. In each iteration, rough clusters and outliers are also identified using our own method called rubikClust.
}
\usage{
ipcaps(bed=NA, rdata=NA, files=NA, label.file=NA, lab.col=1, out, plot.as.pdf=FALSE,
method='mix', missing=NA, covariate=NA, cov.col.first=NA, cov.col.last=NA, 
threshold=0.18, min.fst=0.0008, min.in.group=20, no.plot = FALSE)
}
%- maybe also 'usage' for other objects documented here.
\arguments{
  \item{bed}{
%%     ~~Describe \code{x} here~~
PLINK binary format consists of 3 files; bed, bim, and fam. To generate these files from PLINK, use option \emph{--make-bed}. See more details at: 

http://zzz.bwh.harvard.edu/plink/data.shtml
}
  \item{rdata}{
%%     ~~Describe \code{x} here~~
In case of re-analysis, it is convenient to run IPCAPS using the file rawdata.RData generated by IPCAPS. This file contains a matrix of SNPs (\emph{raw.data}) and a vector of labels (\emph{label}).
}
\item{files}{
%%     ~~Describe \code{x} here~~
IPCAPS supports SNPs encoded as 0, 1 and 2 (dosage encoding). Rows represent SNPs and columns represent subjects. Each column needs to be separated by a space or a tab. A big text file should be divided into smaller files to load faster. For instance, to input 3 files, use as:

\emph{files=c('input1.txt', 'input2.txt', 'input3.txt')}.
}
  \item{label.file}{
%%     ~~Describe \code{x} here~~
An additional useful information (called "labels" in IPCAPS) related subject, for example, geographic location or disease phenotype. These labels (one at a time) are used in displaying the clustering outcome of IPCAPS. A label file must contain at least one column. However, it may contain more than one column in which case each column need to be separated by a space or a tab.
}
  \item{lab.col}{
%%     ~~Describe \code{x} here~~
The label in the label file to be used in the tree-like display of IPCAPS  clustering results.
}
  \item{out}{
%%     ~~Describe \code{x} here~~
To set an absolute path for IPCAPS output. If the specified output directory already exists, result files are saved in sub-directories cluster_out, cluster_out1, cluster_out2, etc.
}
  \item{plot.as.pdf}{
%%     ~~Describe \code{x} here~~
To export plots as PDF. When omitted, plots are saved as PNG.
}
  \item{method}{
%%     ~~Describe \code{x} here~~
The internal clustering method. It can be set to "mix" (rubikClust & mixmod), "mixmod" [2], "clara" [3], "pam" [4], "meanshift" [5], "apcluster" [6], and "hclust" [7]. Default = "mix".
}
 
  \item{missing}{
%%     ~~Describe \code{x} here~~
Symbol used for missing genotypes. Default = \emph{NA}.
}
  \item{covariate}{
%%     ~~Describe \code{x} here~~
A file of covariates; one covariate per column. SNPs can be adjusted for these covariates via regression modeling and residual computation. 
}
  \item{cov.col.first}{
%%     ~~Describe \code{x} here~~
Refer to a covariate file, the first covariate to be considered as confounding variable. 
}
  \item{cov.col.last}{
%%     ~~Describe \code{x} here~~
Refer to a covariate file, the last covariate to be considered as confounding variable. All the variables in between the cov.col.first and cov.col.last will be considered in the adjustment process.
}
  \item{threshold}{
%%     ~~Describe \code{x} here~~
Cutoff value for EigenFit. Possible values range from 0.03 to 0.18. The smaller, the potentially finer the substructure can be. Default = 0.18. 
}
  \item{min.fst}{
%%     ~~Describe \code{x} here~~
Minimum Fst between a pair of subgroups. Default = 0.0008.
}
  \item{min.in.group}{
%%     ~~Describe \code{x} here~~
Minimum number of individuals to constitute a cluster or subgroup. Default = 20.
}
\item{no.plot}{
%%     ~~Describe \code{x} here~~
No plot is generated if this option is TRUE. This option is useful when the system does not support X Windows in the unix based system. Default = FALSE.
}
}
\details{
%%  ~~ If necessary, more details than the description above ~~
The computational time depends on the number of individuals. Consequentially, if large data sets are analyzed, it may be necessary first to split data into smaller files, and then load into computer memory (with parameter 'files'). Internally, the split files are merged to construct a com-prehensive matrix. 
}
\value{
%%  ~Describe the value returned
%%  If it is a LIST, use
%%  \item{comp1 }{Description of 'comp1'}
%%  \item{comp2 }{Description of 'comp2'}
%% ...
The function returns the \emph{list} object containing 2 internal objects;  \emph{output.dir} as class \emph{character} and \emph{cluster} as class \emph{data.frame}. The object \emph{output.dir} stores a result directory. The object \emph{cluster} contains 4 columns, \emph{group}, \emph{node}, \emph{label}, and \emph{row.number}. The column \emph{group} contains the assigned groups from \emph{IPCAPS}.  The column \emph{node} contains node numbers in a tree as illustrated in the \emph{HTML} result files. The column \emph{label} contains the given labels that is set to the parameter \emph{label}. The column \emph{row.number} contains indices to an input data, which is matched to row numbers of input matrix. All clustering result files are saved in one directory (as specified by \emph{out}) containing assigned groups, hierarchical trees of group membership, PC plots, and binary data for further analysis.
}
\references{
%% ~put references to the literature/web site here ~
[1] Intarapanich A, Shaw PJ, Assawamakin A, Wangkumhang P, Ngamphiw C, Chaichoompu K, et al. Iterative pruning PCA improves resolution of highly structured populations. BMC Bioinformatics. 2009;10:382. 

[2] Rmixmod: The R Package of the Model-Based Unsupervised, Supervised, and Semi-Supervised Classification Mixmod Library | Lebret | Journal of Statistical Software [Internet]. [cited 2016 Aug 30]. Available from: https://www.jstatsoft.org/article/view/v067i06

[3] R: Clustering Large Applications [Internet]. [cited 2017 Mar 7]. 

Available from: https://stat.ethz.ch/R-manual/R-devel/library/cluster/html/clara.html

[4] R: Partitioning Around Medoids (PAM) Object [Internet]. [cited 2017 Mar 7]. Available from: https://stat.ethz.ch/R-manual/R-devel/library/cluster/html/pam.object.html

[5] Wang MC and D. MeanShift: Clustering via the Mean Shift Algorithm [Internet]. 2016 [cited 2017 Mar 7]. Available from: https://cran.r-project.org/web/packages/MeanShift/index.html

[6] Bodenhofer U, Palme J, Melkonian C, Kothmeier A. apcluster: Affinity Propagation Clustering [Internet]. 2016 [cited 2017 Mar 7]. 

Available from: https://cran.r-project.org/web/packages/apcluster/index.html

[7] R: Hierarchical Clustering [Internet]. [cited 2017 Mar 7]. Available from: https://stat.ethz.ch/R-manual/R-devel/library/stats/html/hclust.html

}
\author{
%%  ~~who you are~~

}
\note{
%%  ~~further notes~~
}

%% ~Make other sections like Warning with \section{Warning }{....} ~

\seealso{
%% ~~objects to See Also as \code{\link{help}}, ~~~
}
\examples{

# Use the BED format as input
# Importantly, bed file, bim file, and fam file are required
# Use the example files embedded in the package

BED.file <- system.file("extdata","simSNP.bed",package="IPCAPS")
LABEL.file <- system.file("extdata","simSNP_individuals.txt",package="IPCAPS")

my.cluster1 <- ipcaps(bed=BED.file,label.file=LABEL.file,lab.col=2,out=getwd())

table(my.cluster1$cluster$label,my.cluster1$cluster$group)
#             1   2   3   4   5   6
#  outlier4   5   4   1   0   0   0
#  pop1       0   0   0   0 250   0
#  pop2       0   0   0 250   0   0
#  pop3       0   0   0   0   0 250

# Use a text file as input
# Use the example files embedded in the package

text.file <- system.file("extdata","simSNP_data_numMark_rowVar_colInd.txt",
             package="IPCAPS")
LABEL.file <- system.file("extdata","simSNP_individuals.txt",package="IPCAPS")

my.cluster2 <- ipcaps(files=c(text.file),label.file=LABEL.file,lab.col=2,out=getwd())
table(my.cluster2$cluster$label,my.cluster2$cluster$group)
#             1   2   3   4   5   6
#  outlier4   5   4   1   0   0   0
#  pop1       0   0   0   0 250   0
#  pop2       0   0   0 250   0   0
#  pop3       0   0   0   0   0 250

# Use an R Data file as input
# Use the example file embedded in the package

rdata.file <- system.file("data","simSNP.RData",package="IPCAPS")

my.cluster3 <- ipcaps(rdata=rdata.file,out=getwd())
table(my.cluster3$cluster$label,my.cluster3$cluster$group)
#             1   2   3   4   5   6
#  outlier4   5   4   1   0   0   0
#  pop1       0   0   0   0 250   0
#  pop2       0   0   0 250   0   0
#  pop3       0   0   0   0   0 250
}
% Add one or more standard keywords, see file 'KEYWORDS' in the
% R documentation directory.

